htlcswitch/switch.go

```diff
@@ -361,34 +361,48 @@ func (s *Switch) SendHTLC(firstHop lnwire.ShortChannelID,
 	preimageChan := make(chan [sha256.Size]byte, 1)
 	errChan := make(chan error, 1)
 
-	// Before sending, double check that we don't already have 1) an
-	// in-flight payment to this payment hash, or 2) a complete payment for
-	// the same hash.
-	if err := s.control.ClearForTakeoff(htlc); err != nil {
+	paymentID, err := s.paymentSequencer.NextID()
+	if err != nil {
 		errChan <- err
 		return preimageChan, errChan
 	}
 
-	// Create payment and add to the map of payment in order later to be
-	// able to retrieve it and return response to the user.
-	payment := &pendingPayment{
-		err:          errChan,
-		preimage:     preimageChan,
-		paymentHash:  htlc.PaymentHash,
-		amount:       htlc.Amount,
-		deobfuscator: deobfuscator,
-	}
+	// Before sending, double check that we don't already have 1) an
+	// in-flight payment to this payment hash, or 2) a complete payment for
+	// the same hash.
+	s.pendingMutex.Lock()
```

Comment:
do we need to hold the lock around `ClearForTakeoff`? this limits it to single writer, and we lose out on all benefits of using `Batch`

Reply:
The reason I did this was to handle the case where `ClearForTakeoff` is called concurrently with a HTLC for the same payment has gets settled. Without the lock we risk getting `ErrPaymentInFlight` returned, but before we add the pending payment to the map the HTLC is settled.

I do see the lost benefit of batch... Time to bring out the `multimutex`? 😁

Reply:
Used a multimutex, was the only way to really ensure the `CircuitMap` and `ControlTower` would always stay in sync.